### PR TITLE
Renovate build dependencies

### PR DESCRIPTION
**Pull request by kyessenov**
- upgrade to bazel 0.5.2, latest rules_go
- remove rules_docker in favor of Dockerfiles
- use static go build
- use scratch images for go docker images
- move iptables to the init container base image
- use tmpfs for temporary files in the sidecar container and make container read only

### CODE REVIEW

**Comment by kyessenov** [comment]
Fixes #904

**Comment by kyessenov** [comment]
Fixes #952

**Comment by kyessenov** [comment]
@sebastienvas please update bazel version

File: platform/kube/inject/testdata/auth.non-default-service-account.yaml.injected [hunk 1]
```diff
@@ -60,6 +63,9 @@ spec:
       - configMap:
           name: istio
         name: istio-config
+      - emptyDir:
+          medium: Memory
```
Comment (rshriram) [inline platform/kube/inject/testdata/auth.non-default-service-account.yaml.injected]: What does this mean, emptyDir medium memory?

Comment (kyessenov) [inline platform/kube/inject/testdata/auth.non-default-service-account.yaml.injected]: This is to mount the envoy config directory as tmpfs (an in-memory file system).

File: WORKSPACE [hunk 2]
```diff
@@ -259,138 +259,126 @@ new_go_repository(
...
 
-DEBUG_BASE_IMAGE_SHA = "3350805cd019192b1a71fcdcc786ac4a8500e0d4f96c30e5e4c8ff43b9208195"
+ISTIO_IPTABLES_TAG = "20170725-5a9fbbf"
 
-http_file(
-    name = "ubuntu_xenial_debug",
-    sha256 = DEBUG_BASE_IMAGE_SHA,
-    url = "https://storage.googleapis.com/istio-build/pilot/ubuntu_xenial_debug-" + DEBUG_BASE_IMAGE_SHA + ".tar.gz",
+docker_pull(
+    name = "istio_iptables",
```
Comment (kyessenov) [inline WORKSPACE]: The init image is called gcr.io/istio-testing/init:_. I'm fine renaming that one to istio_init. This is the base that just contains iptables + xenial.

**Comment by kyessenov** [comment]
We use bazel docker just to layer pilot binaries on top. Bazel can do pull/push these days.

**Comment by ldemailly** [comment]
this looks great and also a lot of work, thanks. any volunteer for the other repos ?
(and as we discussed we should find a way to automate and share some of that more)

**Comment by sebastienvas** [comment]
build pilot/presubmit

**Comment by sebastienvas** [comment]
@kuat sorry for the delay. We changed the prowbazel image as well and I did not want conflicts. I only updated the jenkins image. I am working on  update prowbazel now.

**Comment by kyessenov** [comment]
This revision completely removes bazel docker rules.

**Comment by sebastienvas** [comment]
Please discard prow for now. We are fixing it.

**Pull request merged**